File: sdk/go/coin/transactions.go
package coin

import (
	"context"
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"time"
)

var outTradeNoPattern = regexp.MustCompile(`^[A-Za-z0-9_-]{1,64}$`)

// TransactionsAPI contains transaction submit and query APIs.
type TransactionsAPI struct {
	client *Client
}

func (t *TransactionsAPI) Credit(ctx context.Context, req CreditRequest) (TxnSubmitResponse, error) {
	var err error
	req.OutTradeNo, err = normalizeOutTradeNoAndAmount(req.OutTradeNo, req.Amount)
	if err != nil {
		return TxnSubmitResponse{}, err
	}
	if strings.TrimSpace(req.CreditAccountNo) == "" && strings.TrimSpace(req.UserID) == "" {
		return TxnSubmitResponse{}, fmt.Errorf("credit_account_no or user_id is required")
	}
	if req.ExpireInDays < 0 {
		return TxnSubmitResponse{}, fmt.Errorf("expire_in_days must be >= 0")
	}
	var out TxnSubmitResponse
	if err := t.client.do(ctx, "POST", "/api/v1/transactions/credit", nil, req, &out); err != nil {
		return TxnSubmitResponse{}, err
	}
	return out, nil
}

func (t *TransactionsAPI) Debit(ctx context.Context, req DebitRequest) (TxnSubmitResponse, error) {
	var err error
	req.OutTradeNo, err = normalizeOutTradeNoAndAmount(req.OutTradeNo, req.Amount)
	if err != nil {
		return TxnSubmitResponse{}, err
	}
	if strings.TrimSpace(req.DebitAccountNo) == "" && strings.TrimSpace(req.DebitOutUserID) == "" {
		return TxnSubmitResponse{}, fmt.Errorf("debit_account_no or debit_out_user_id is required")
	}
	var out TxnSubmitResponse
	if err := t.client.do(ctx, "POST", "/api/v1/transactions/debit", nil, req, &out); err != nil {
		return TxnSubmitResponse{}, err
	}
	return out, nil
}

func (t *TransactionsAPI) Transfer(ctx context.Context, req TransferRequest) (TxnSubmitResponse, error) {
	var err error
	req.OutTradeNo, err = normalizeOutTradeNoAndAmount(req.OutTradeNo, req.Amount)
	if err != nil {
		return TxnSubmitResponse{}, err
	}
	if strings.TrimSpace(req.FromAccountNo) == "" && strings.TrimSpace(req.FromOutUserID) == "" {
		return TxnSubmitResponse{}, fmt.Errorf("from_account_no or from_out_user_id is required")
	}
	if strings.TrimSpace(req.ToAccountNo) == "" && strings.TrimSpace(req.ToOutUserID) == "" {
		return TxnSubmitResponse{}, fmt.Errorf("to_account_no or to_out_user_id is required")
	}
	if req.ToExpireInDays < 0 {
		return TxnSubmitResponse{}, fmt.Errorf("to_expire_in_days must be >= 0")
	}
	var out TxnSubmitResponse
	if err := t.client.do(ctx, "POST", "/api/v1/transactions/transfer", nil, req, &out); err != nil {
		return TxnSubmitResponse{}, err
	}
	return out, nil
}

func (t *TransactionsAPI) Refund(ctx context.Context, req RefundRequest) (TxnSubmitResponse, error) {
	var err error
	req.OutTradeNo, err = normalizeOutTradeNoAndAmount(req.OutTradeNo, req.Amount)
	if err != nil {
		return TxnSubmitResponse{}, err
	}
	req.RefundOfTxnNo = strings.TrimSpace(req.RefundOfTxnNo)
	if req.RefundOfTxnNo == "" {
		return TxnSubmitResponse{}, fmt.Errorf("refund_of_txn_no is required")
	}
	var out TxnSubmitResponse
	if err := t.client.do(ctx, "POST", "/api/v1/transactions/refund", nil, req, &out); err != nil {
		return TxnSubmitResponse{}, err
	}
	return out, nil
}

func (t *TransactionsAPI) GetByTxnNo(ctx context.Context, txnNo string) (Txn, error) {
	txnNo = strings.TrimSpace(txnNo)
	if txnNo == "" {
		return Txn{}, fmt.Errorf("txn_no is required")
	}
	var out Txn
	if err := t.client.do(ctx, "GET", "/api/v1/transactions/"+url.PathEscape(txnNo), nil, nil, &out); err != nil {
		return Txn{}, err
	}
	return out, nil
}

func (t *TransactionsAPI) GetByOutTradeNo(ctx context.Context, outTradeNo string) (Txn, error) {
	outTradeNo = strings.TrimSpace(outTradeNo)
	if outTradeNo == "" {
		return Txn{}, fmt.Errorf("out_trade_no is required")
	}
	var out Txn
	if err := t.client.do(ctx, "GET", "/api/v1/transactions/by-out-trade-no/"+url.PathEscape(outTradeNo), nil, nil, &out); err != nil {
		return Txn{}, err
	}
	return out, nil
}

func (t *TransactionsAPI) List(ctx context.Context, req ListTransactionsRequest) (ListTransactionsResponse, error) {
	query := url.Values{}
	if req.StartTime != nil {
		query.Set("start_time", req.StartTime.UTC().Format(time.RFC3339))
	}
	if req.EndTime != nil {
		query.Set("end_time", req.EndTime.UTC().Format(time.RFC3339))
	}
	if v := strings.TrimSpace(req.Status); v != "" {
		query.Set("status", strings.ToUpper(v))
	}
	if v := strings.TrimSpace(req.TransferScene); v != "" {
		query.Set("transfer_scene", strings.ToUpper(v))
	}
	if v := strings.TrimSpace(req.OutUserID); v != "" {
		query.Set("out_user_id", v)
	}
	if req.PageSize > 0 {
		query.Set("page_size", fmt.Sprintf("%d", req.PageSize))
	}
	if v := strings.TrimSpace(req.PageToken); v != "" {
		query.Set("page_token", v)
	}
	var out ListTransactionsResponse
	if err := t.client.do(ctx, "GET", "/api/v1/transactions", query, nil, &out); err != nil {
		return ListTransactionsResponse{}, err
	}
	if out.Items == nil {
		out.Items = make([]Txn, 0)
	}
	return out, nil
}

func (t *TransactionsAPI) ListAccountChangeLogs(ctx context.Context, accountNo string, req ListAccountChangeLogsRequest) (ListAccountChangeLogsResponse, error) {
	accountNo = strings.TrimSpace(accountNo)
	if accountNo == "" {
		return ListAccountChangeLogsResponse{}, fmt.Errorf("account_no is required")
	}

	query := url.Values{}
	if req.PageSize > 0 {
		query.Set("page_size", fmt.Sprintf("%d", req.PageSize))
	}
	if v := strings.TrimSpace(req.PageToken); v != "" {
		query.Set("page_token", v)
	}

	var out ListAccountChangeLogsResponse
	if err := t.client.do(ctx, "GET", "/api/v1/accounts/"+url.PathEscape(accountNo)+"/change-logs", query, nil, &out); err != nil {
		return ListAccountChangeLogsResponse{}, err
	}
	if out.Items == nil {
		out.Items = make([]AccountChangeLog, 0)
	}
	return out, nil
}

func normalizeOutTradeNoAndAmount(outTradeNo string, amount int64) (string, error) {
	trimmed := strings.TrimSpace(outTradeNo)
	if trimmed == "" {
		return "", fmt.Errorf("out_trade_no is required")
	}
	if !outTradeNoPattern.MatchString(trimmed) {
		return "", fmt.Errorf("invalid out_trade_no")
	}
	if amount <= 0 {
		return "", fmt.Errorf("amount must be > 0")
	}
	return trimmed, nil
}

File: sdk/ts/coin/src/transactions.ts
import type { CoinClient } from "./client.js";
import type {
  CreditRequest,
  DebitRequest,
  TransferRequest,
  RefundRequest,
  ListTransactionsRequest,
  TxnSubmitResponse,
  Txn,
  ListTransactionsResponse,
  ListAccountChangeLogsRequest,
  ListAccountChangeLogsResponse,
} from "./types.js";

const OUT_TRADE_NO_PATTERN = /^[A-Za-z0-9_-]{1,64}$/;

export class TransactionsAPI {
  /** @internal */
  private readonly client: CoinClient;

  /** @internal */
  constructor(client: CoinClient) {
    this.client = client;
  }

  async credit(req: CreditRequest): Promise<TxnSubmitResponse> {
    req = { ...req, out_trade_no: validateOutTradeNoAndAmount(req.out_trade_no, req.amount) };
    if (!(req.credit_account_no ?? "").trim() && !(req.user_id ?? "").trim()) {
      throw new Error("credit_account_no or user_id is required");
    }
    if ((req.expire_in_days ?? 0) < 0) {
      throw new Error("expire_in_days must be >= 0");
    }
    return this.client._do<TxnSubmitResponse>("POST", "/api/v1/transactions/credit", undefined, req);
  }

  async debit(req: DebitRequest): Promise<TxnSubmitResponse> {
    req = { ...req, out_trade_no: validateOutTradeNoAndAmount(req.out_trade_no, req.amount) };
    if (!(req.debit_account_no ?? "").trim() && !(req.debit_out_user_id ?? "").trim()) {
      throw new Error("debit_account_no or debit_out_user_id is required");
    }
    return this.client._do<TxnSubmitResponse>("POST", "/api/v1/transactions/debit", undefined, req);
  }

  async transfer(req: TransferRequest): Promise<TxnSubmitResponse> {
    req = { ...req, out_trade_no: validateOutTradeNoAndAmount(req.out_trade_no, req.amount) };
    if (!(req.from_account_no ?? "").trim() && !(req.from_out_user_id ?? "").trim()) {
      throw new Error("from_account_no or from_out_user_id is required");
    }
    if (!(req.to_account_no ?? "").trim() && !(req.to_out_user_id ?? "").trim()) {
      throw new Error("to_account_no or to_out_user_id is required");
    }
    if ((req.to_expire_in_days ?? 0) < 0) {
      throw new Error("to_expire_in_days must be >= 0");
    }
    return this.client._do<TxnSubmitResponse>("POST", "/api/v1/transactions/transfer", undefined, req);
  }

  async refund(req: RefundRequest): Promise<TxnSubmitResponse> {
    req = { ...req, out_trade_no: validateOutTradeNoAndAmount(req.out_trade_no, req.amount) };
    const refundOfTxnNo = (req.refund_of_txn_no ?? "").trim();
    if (!refundOfTxnNo) throw new Error("refund_of_txn_no is required");
    req = { ...req, refund_of_txn_no: refundOfTxnNo };
    return this.client._do<TxnSubmitResponse>("POST", "/api/v1/transactions/refund", undefined, req);
  }

  async getByTxnNo(txnNo: string): Promise<Txn> {
    txnNo = (txnNo ?? "").trim();
    if (!txnNo) throw new Error("txn_no is required");
    return this.client._do<Txn>("GET", `/api/v1/transactions/${encodeURIComponent(txnNo)}`);
  }

  async getByOutTradeNo(outTradeNo: string): Promise<Txn> {
    outTradeNo = (outTradeNo ?? "").trim();
    if (!outTradeNo) throw new Error("out_trade_no is required");
    return this.client._do<Txn>("GET", `/api/v1/transactions/by-out-trade-no/${encodeURIComponent(outTradeNo)}`);
  }

  async list(req: ListTransactionsRequest): Promise<ListTransactionsResponse> {
    const query: Record<string, string> = {};
    if (req.startTime) query.start_time = req.startTime.toISOString();
    if (req.endTime) query.end_time = req.endTime.toISOString();
    if ((req.status ?? "").trim()) query.status = req.status!.trim().toUpperCase();
    if ((req.transferScene ?? "").trim()) query.transfer_scene = req.transferScene!.trim().toUpperCase();
    if ((req.outUserID ?? "").trim()) query.out_user_id = req.outUserID!.trim();
    if (req.pageSize && req.pageSize > 0) query.page_size = String(req.pageSize);
    if ((req.pageToken ?? "").trim()) query.page_token = req.pageToken!.trim();

    const resp = await this.client._do<ListTransactionsResponse>(
      "GET",
      "/api/v1/transactions",
      Object.keys(query).length > 0 ? query : undefined,
    );
    return { ...resp, items: resp.items ?? [] };
  }

  async listAccountChangeLogs(accountNo: string, req: ListAccountChangeLogsRequest = {}): Promise<ListAccountChangeLogsResponse> {
    accountNo = (accountNo ?? "").trim();
    if (!accountNo) throw new Error("account_no is required");

    const query: Record<string, string> = {};
    if (req.pageSize && req.pageSize > 0) query.page_size = String(req.pageSize);
    if ((req.pageToken ?? "").trim()) query.page_token = req.pageToken!.trim();

    const resp = await this.client._do<ListAccountChangeLogsResponse>(
      "GET",
      `/api/v1/accounts/${encodeURIComponent(accountNo)}/change-logs`,
      Object.keys(query).length > 0 ? query : undefined,
    );
    return { ...resp, items: resp.items ?? [] };
  }
}

function validateOutTradeNoAndAmount(outTradeNo: string, amount: number): string {
  const trimmed = (outTradeNo ?? "").trim();
  if (!trimmed) throw new Error("out_trade_no is required");
  if (!OUT_TRADE_NO_PATTERN.test(trimmed)) throw new Error("invalid out_trade_no");
  if (!amount || amount <= 0) throw new Error("amount must be > 0");
  return trimmed;
}
